tutorials/how-to-implement-rag-generativeapis/index.mdx

```diff
@@ -0,0 +1,357 @@
+page_iterator = paginator.paginate(Bucket=os.getenv("SCW_BUCKET_NAME", ""))
+```
+
+In this code sample we:
+- Set up a Boto3 session: We initialize a Boto3 session, which is the AWS SDK for Python, fully compatible with Scaleway Object Storage. This session manages configuration, including credentials and settings, that Boto3 uses for API requests.
+- Create an S3 client: We establish an S3 client to interact with the Scaleway Object Storage service.
+- Set up pagination for listing objects: We prepare pagination to handle potentially large lists of objects efficiently.
+- Iterate through the bucket: This initiates the pagination process, allowing us to list all objects within the specified Scaleway Object bucket seamlessly.
+
+### Iterate through metadata
+
+Next, we will iterate through the metadata to determine if each object has already been embedded. If an object hasn’t been processed yet, we will embed it and load it into the database.
+
+```python
+# rag.py
+
+text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=0, add_start_index=True, length_function=len, is_separator_regex=False)
+for page in page_iterator:
+    for obj in page.get('Contents', []):
+        cur.execute("SELECT object_key FROM object_loaded WHERE object_key = %s", (obj['Key'],))
+        response = cur.fetchone()
+        if response is None:
+            file_loader = S3FileLoader(
+                    bucket=os.getenv("SCW_BUCKET_NAME", ""),
+                    key=obj['Key'],
+                    endpoint_url=os.getenv("SCW_BUCKET_ENDPOINT", ""),
+                    aws_access_key_id=os.getenv("SCW_ACCESS_KEY", ""),
+                    aws_secret_access_key=os.getenv("SCW_API_KEY", "")
+                )
+            file_to_load = file_loader.load()
+            cur.execute("INSERT INTO object_loaded (object_key) VALUES (%s)", (obj['Key'],))
+            chunks = text_splitter.split_text(file_to_load[0].page_content)
+            try:
+                embeddings_list = [embeddings.embed_query(chunk) for chunk in chunks]
+                vector_store.add_embeddings(chunks, embeddings_list)
+                cur.execute("INSERT INTO object_loaded (object_key) VALUES (%s)", (obj['Key'],))
+            except Exception as e:
+                logger.error(f"An error occurred: {e}")
+
+conn.commit()
+```
+
+- S3FileLoader: The S3FileLoader loads each file individually from your ***Scaleway Object Storage bucket*** using the file's object_key (extracted from the file's metadata). It ensures that only the specific file is loaded from the bucket, minimizing the amount of data being retrieved at any given time.
+- RecursiveCharacterTextSplitter: The RecursiveCharacterTextSplitter breaks each document into smaller chunks of text. This is crucial because embeddings models, like those used in Retrieval-Augmented Generation (RAG), typically have a limited context window (the number of tokens they can process at once).
```

Comment:
```suggestion
- RecursiveCharacterTextSplitter: the RecursiveCharacterTextSplitter breaks each document into smaller chunks of text. This is crucial, because embeddings models, like those used in Retrieval-Augmented Generation (RAG), typically have a limited context window (the number of tokens they can process at once).
```